src/commands/run.rs

```diff
@@ -423,22 +426,32 @@ impl RunCommand {
         // we don't actually read the whole file into memory. Note that this
         // behavior is disabled by default, though, because it's not safe to
         // pass arbitrary user input to this command with `--allow-precompiled`
-        let mut file =
-            File::open(path).with_context(|| format!("failed to open: {}", path.display()))?;
-        let mut magic = [0; 4];
-        if let Ok(()) = file.read_exact(&mut magic) {
-            if &magic == b"\x7fELF" {
-                if self.allow_precompiled {
-                    return unsafe { Module::deserialize_file(engine, path) };
-                }
-                bail!(
-                    "cannot load precompiled module `{}` unless --allow-precompiled is passed",
-                    path.display()
-                )
+
+        // Try to mmap the module file.
+        // If it fails(it wouldn't be a regular file), allocate a fixed mmap buffer
+        // to store the whole bytes in a memory.
+        let mmap_vec = MmapVec::from_file(path).or_else(|_| {
+            // Mmap-ing this much would not eat system memory thanks for paging
+            let mut buf = Vec::new();
+            let mut f = File::open(path)?;
+            f.read_to_end(&mut buf)?;
+            let mmap_vec = MmapVec::from_slice(&buf)?;
+            Ok::<_, anyhow::Error>(mmap_vec)
+        })?;
```

Comment:
This introduces new copies of the data, slowing down startup. Previously `Module::from_file` would mmap the file once and then use this mmap'ed version directly. Now it will mmap it once (or read to a vec and then copy it to a new memory region in `MmapVec::from_slice`) and then copy it to a new memory region again in `Module::new` (through `MmapVec::from_slice`).

Reply:
Definitely right. To be honest, I thought directly writing Mmap/MmapVec with looped `f.read()` is more efficient but feeled like a premature optimization, so chose this implementation for brevity. (The main reason for this is the File::read_to_end is not compatible with MmapVec.)

If startup performance matters for piped inputs to, I'll fix this into like
```rust
loop {
    let buf = [0u8; 4<<10];
    match f.read(&mut buf) {
        Ok(0) => break,
        Ok(n) => /* copy `n` bytes to Mmap and increment the offset */ todo!(),
        Err(e) => return Err(e),
    }
}
```

But again I'm not sure this is worth it.

Reply:
This will still have an extra copy in case the mmap succeeded.

Reply:
OK I didn't understand what you were saying. 😅 Now I see.
I'm afraid that this is not able to be trivially fixed, as there's no public API on `Module` to pass MmapVec directly into a constructor?

`from_binary` seems to be the 'generic' catch-all for many Module constructors but I don't have an idea what it does...

Reply:
Nevermind, you said about the case when the mmap suceeded, so trying to mmap and if it succeeds then discarding the mmap and use Module::from_path is OK. Is that you're thinking?

Reply:
That may work.